### DISCUSSION TITLE
Preventing Data Loss

### DISCUSSION THREAD

**Discussion by Xevion** [twpayne/chezmoi#3986] [Q&A] [answered]
Hi,

I was curious if Chezmoi or it's users have methods for making sure I don't `apply` without being super sure data loss won't occur.

The wider I expand my dotfiles config footprint, the more likely I am going to be working with a program or config that either changes it config autonomously, or I will change it without remembering to use Chezmoi.

Of course, this isn't problematic on it's own since Chezmoi can merge/pull these changes in, but I find it likely that I will:

1. Have a destination file that's edited.
2. Work on something unrelated to it in the source.
3. Use `chezmoi apply`, not using `--interactive` or `chezmoi diff` prior.
4. Destroy data in a config I did not mean to - and without me knowing immediately.
5. Pain.

So far, I found that the only real way is to make sure I use `chezmoi edit -a` or `chezmoi diff` or `chezmoi apply --interactive` prior. But for obvious reasons, I'm going to forget and endure data loss, and being ultra careful at all times isn't exactly a 'solution' in my mind.

More or less, I wish Chezmoi had options to provide data loss mitigation or warnings by default. I'm hoping the community may offer suggestions from their experience, or point me in the right direction to using existing solutions that I am unaware of.

**Comment by twpayne** [comment, marked as answer]
chezmoi already does this.

You can always use `chezmoi diff` to see exactly what changes chezmoi will make to your system.

`chezmoi apply` will warn you if a file managed by chezmoi has been changed since chezmoi last wrote it.

`chezmoi managed` and `chezmoi unmanaged` tell you exactly which files are managed by chezmoi.

chezmoi encourages you to keep your dotfiles in version control, so you can always roll back to a previous version.

Of course, in the case of accidentally overwriting a file, you can always retrieve the file from your backup system. chezmoi is a dotfile manager, not a backup system.

> More or less, I wish Chezmoi had options to provide data loss mitigation or warnings by default. I'm hoping the community may offer suggestions from their experience, or point me in the right direction to using existing solutions that I am unaware of.

What, exactly, do you think chezmoi should do that it doesn't do already?

**Reply by Xevion** [reply]
> `chezmoi apply` will warn you if a file managed by chezmoi has been changed since chezmoi last wrote it.

I was unaware of this. Damn. :eyes:
